pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionCommon.java

```diff
@@ -383,4 +377,112 @@ public static double roundDecimal(double value, int places) {
+                throw new IllegalArgumentException(String.format("Failed to extract %s class from archive",
+                        componentType.toString().toLowerCase()), e);
+            }
+
+            try {
+                narClassLoader.loadClass(connectorClassName);
+                classLoader = narClassLoader;
+            } catch (ClassNotFoundException | NoClassDefFoundError e) {
+                throw new IllegalArgumentException(
+                        String.format("%s class %s must be in class path", capFirstLetter(componentType), connectorClassName), e);
+            }
+
+        } else {
+            // if connector class name is provided, we need to try to load it as a JAR and as a NAR.
+            if (jarClassLoader != null) {
+                try {
+                    jarClassLoader.loadClass(connectorClassName);
+                    classLoader = jarClassLoader;
+                } catch (ClassNotFoundException | NoClassDefFoundError e) {
+                    // class not found in JAR try loading as a NAR and searching for the class
+                    if (narClassLoader != null) {
+
+                        try {
+                            narClassLoader.loadClass(connectorClassName);
+                            classLoader = narClassLoader;
+                        } catch (ClassNotFoundException | NoClassDefFoundError e1) {
+                            throw new IllegalArgumentException(
+                                    String.format("%s class %s must be in class path",
+                                            capFirstLetter(componentType), connectorClassName), e1);
+                        }
+                    } else {
+                        throw new IllegalArgumentException(
+                                String.format("%s class %s must be in class path", capFirstLetter(componentType),
+                                        connectorClassName), e);
+                    }
+                }
+            } else if (narClassLoader != null) {
+                try {
+                    narClassLoader.loadClass(connectorClassName);
+                    classLoader = narClassLoader;
+                } catch (ClassNotFoundException | NoClassDefFoundError e1) {
+                    throw new IllegalArgumentException(
+                            String.format("%s class %s must be in class path",
+                                    capFirstLetter(componentType), connectorClassName), e1);
+                }
+            } else {
+                StringBuilder errorMsg = new StringBuilder(capFirstLetter(componentType)
+                        + " package does not have the correct format."
+                        + " Pulsar cannot determine if the package is a NAR package or JAR package.");
+
+                if (jarClassLoaderException != null) {
+                    errorMsg.append("Attempts to load it as a JAR package produced error: " + jarClassLoaderException.getMessage());
+                }
+
```

Comment:
do you want to add some linebreak or delimiter between these two error message for better reading?

Reply:
I can add a space